relay-metrics/src/aggregation.rs

```diff
@@ -69,6 +70,365 @@ impl GaugeValue {
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use relay_metrics::dist;
+    ///
+    /// let dist = dist![2.0, 1.0, 3.0, 2.0];
+    ///
+    /// let mut iter = dist.iter_values();
+    /// assert_eq!(iter.next(), Some(1.0));
+    /// assert_eq!(iter.next(), Some(2.0));
+    /// assert_eq!(iter.next(), Some(2.0));
+    /// assert_eq!(iter.next(), Some(3.0));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    pub fn iter_values(&self) -> DistributionValuesIter<'_> {
+        DistributionValuesIter {
+            inner: self.iter(),
+            current: 0f64,
+            remaining: 0,
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a DistributionValue {
+    type Item = (f64, u32);
+    type IntoIter = DistributionIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl fmt::Debug for DistributionValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl Extend<f64> for DistributionValue {
+    fn extend<T: IntoIterator<Item = f64>>(&mut self, iter: T) {
+        for value in iter.into_iter() {
+            self.insert(value);
+        }
+    }
+}
+
+impl Extend<(f64, u32)> for DistributionValue {
+    fn extend<T: IntoIterator<Item = (f64, u32)>>(&mut self, iter: T) {
+        for (value, count) in iter.into_iter() {
+            self.insert_multi(value, count);
+        }
+    }
+}
+
+impl Serialize for DistributionValue {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.collect_seq(self.iter_values())
+    }
+}
+
+impl<'de> Deserialize<'de> for DistributionValue {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct DistributionVisitor;
+
+        impl<'d> serde::de::Visitor<'d> for DistributionVisitor {
+            type Value = DistributionValue;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(formatter, "a list of floating point values")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'d>,
+            {
+                let mut distribution = DistributionValue::new();
+
+                while let Some(value) = seq.next_element()? {
+                    distribution.insert(value);
+                }
+
+                Ok(distribution)
+            }
+        }
+
+        deserializer.deserialize_seq(DistributionVisitor)
+    }
+}
+
+/// An iterator over distribution entries in a [`DistributionValue`].
+///
+/// This struct is created by the [`iter`](DistributionValue::iter) method on
+/// `DistributionValue`. See its documentation for more.
+#[derive(Clone)]
+pub struct DistributionIter<'a> {
```

Comment:
Should we declare the iterators before `DistributionValue`?

Reply:
Thanks for the review, @untitaker! 

> * if singles ever gets too large we may want to consider rounding values such that we can have more duplicates

We can consider of lowering precision, although I'm not sure what reasonable thresholds should be. At the time, we're committed to keeping full precision in Relay.

> * singles could be some sort of bloom filter contraption

This is the form of lossy aggregation we currently leave to storage. Potentially, we add this to Relay in future, too, although then using the same sketching algorithm also used in storage.

> may want to add some tests for NaN since those can have different bit reprs and I am not sure if float_ord gets equality right.

Good point, reporting `NaN` doesn't make sense on any of our metrics. @jjbayer this is something we can catch at an earlier stage before merging into the aggregation.